Playwright: restructure tests to remove non-determinism

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR proposes to remove several tests and restructure others in order to eliminate non-determinism.

Key Changes:
- removal of `Comment and like on an existing post` suite.
- combination of `Like a new post` and `Like post as another user` suite into one flow.

Background Details:

When initially migrating the tests from Selenium to Playwright, I made an effort to expand the scope of features being tested. While this was a good initiative, some tests ended up being non-deterministic leading to flaky tests.

For instance, the `Comment and like on an existing post` suite relies on the test account:
- having at least 1 published post, and;
- the published post is not private or otherwise protected, and;
- the post supports commenting.

These factors introduce a level of uncertainty into specs. Did the test fail because of a conflict? Or is the failure a _real_ regression?

I feel the best course of action at this time is to restructure specs to test a combined flow where the having multiple flows is not warranted to save both testing resources and time but also make subsequent portions of the flow dependent on earlier steps successfully completing.

#### Testing instructions

- [ ] TeamCity should not show failures relating to the Likes tests.

Related to #

worldomonation committed:
Ensure GutenbergEditorPage.publish has reliable navigation to the published post.
- return the publishedURL by default, to be used in assertion.
- modify selector to refer to the post publish panel's View Post/Page button, instead of the snackbar toast.

Skip the `like an existing post` due to non-determinism this causes.

worldomonation committed:
Unskip `Like an existing post` test.

worldomonation committed:
Remove nondeterministic tests:
- Comment and like on an exising post
- Like an existing post

These tests are problematic because it does not create a post prior to the like/comment process, so it cannot be guaranteed that target posts are in a good state to accept these interactions.

Add assertions after publishing post.

worldomonation committed:
Merge the Like: Logged out user test with the main Like: Post spec file.

- use the same post that was previously created in the test.

worldomonation committed:
The destruction one missing `await` statement can do

worldomonation committed:
Add comments.

worldomonation committed:
Remove commented out code and additional comments.

worldomonation commented:
Oops, still in draft - need to clean up a few things.

worldomonation committed:
Clean up commented code.

worldomonation commented on test/e2e/specs/specs-playwright/wp-likes__post-spec.js:
```diff
@@ -57,33 +62,20 @@ describe( DataHelper.createSuiteTitle( 'Likes (Post)' ), function () {
+			// Clicking the Like button will bring up a new popup.
+			// `loginFromPopup` will observe for a popup event, grab the new popup and
+			// execute the login process on that page.
+			await Promise.all( [ loginFlow.logInFromPopup(), publishedPostPage.likePost() ] );
```
This step is copied entirely over from the now-deleted file `test/e2e/specs/specs-playwright/wp-likes__logged-out-spec.js`.

dpasque commented on packages/calypso-e2e/src/browser-manager.ts:
```diff
@@ -114,11 +114,31 @@ export async function close(): Promise< void > {
+	const browserContext = page.context();
+	const currentURL = page.url();
+
+	// Navigate to the WordPress.com base URL.
+	await page.goto( config.get( 'calypsoBaseURL' ) );
+	// Clear the GDPR banner cookie.
+	await page.evaluate( 'window.document.cookie = "sensitive_pixel_option=no;";' );
+	// Clear additional cookies.
+	await page.evaluate(
+		'window.document.cookie = "sensitive_pixel_option=no;domain=.wordpress.com;SameSite=None;Secure"'
+	);
+	// Clear cookies to remote login.
```
Could you explain a little more what's going on in these lines?

From what I can tell, the goal is to wipe cookies (which is what we do in the end with `browserContext.clearCookies()`). It looks like the goal in these previous steps is mostly to wipe `localstorage` on `r-login.wordpress.com`. 

Is there any reason we can't trim this down to:
```typescript
	await page.goto( 'https://r-login.wordpress.com/' );
	// Clear local storage.
	await page.evaluate( 'localStorage.clear();' );
	// Lastly, clear the cookies using built-in method.
	await browserContext.clearCookies();
	// Previous steps navigated page away from target page. Return page to the original URL.
	await page.goto( currentURL );
```

worldomonation replied:
I copied this behavior from the existing Selenium suite. This behavior can be found [here](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/lib/driver-manager.js#L353-L357), [here](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/lib/driver-manager.js#L304-L313) and [here](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/lib/driver-manager.js#L319-L343).

I think it may be worth trying to see if what you suggested suffices though.

worldomonation replied:
You were correct!

dpasque commented on packages/calypso-e2e/src/browser-manager.ts:
```diff
@@ -114,11 +114,31 @@ export async function close(): Promise< void > {
  * @param {Page} page Object representing a page launched by Playwright.
  * @returns {Promise<void>} No return value.
  */
 export async function clearCookies( page: Page ): Promise< void > {
```
Optional: but what do you think about instead adding a wrapper function around this function, since this function now technically does more than just clearing cookies?

So this function, `clearCookies` can stay just clearing cookies, and then you make a wrapper function called `resetForNewUserLogin` that does the local storage wipe and calls into `clearCookies`?

I just don't want anyone to call this and miss the `localstorage` side effect if they don't know about it (and don't pay attention to the JSDoc header)

worldomonation replied:
I think it may be better to rename the method instead, since the `clearCookies` call is a one-liner (assuming the context is available) and I don't anticipate a situation that may require one but not the other.

What do you think about a name like `clearAuthenticationState` to be more in line with the original Selenium call (`ensureNotLoggedIntoSite`)?

dpasque replied:
Oh, I like that! That name sounds nice!!

dpasque reviewed (changes requested):
Have a couple questions on the browser cookie clearing!

The script updates look really nice - I 100% agree that it is much better to have fewer, more deterministic tests. These flows look really nice, simple, and stable!

worldomonation committed:
Try Dan's suggestions:
- shortened steps to clear local authentication state.
- renamed `clearCookies` to `clearAuthenticationState`

dpasque reviewed (approved):
I think that new version of the function looks really clean and nice - love it!

LGTM! 🚢
